### PR TITLE
Store layers testing and concurrency

### PR DESCRIPTION
**Pull request by bcebere**
## Description
Tests:
- [x] Add unit tests and docstrings for `action_store`.
- [x] Add unit tests and docstrings for `dict_document_store`.
- [x] Add unit tests and docstrings for `kv_document_store`.
- [x] Add unit tests and docstrings for `mongo_document_store`.
- [x] Add unit tests and docstrings for `sqlite_document_store`.
- [x] Add unit tests and docstrings for `queue_stash`.
- [x] Add multithreaded tests for all the stores. 
- [x] Add sqlite3 WAL (thanks @koenvanderveen )
- [x] add distributed locks(no integration yet)

Bugfixing:
- [x]  Fix `action_store` crash on `delete`.
- [x]  Change `init_store` calls to return Result instead of an exception.
- [x]  Extend configuration options for the MongoClient
- [x]  Better error handling for the Mongo store.
- [x] Mongo update functional. 

Observations:
 - Mongo stores cannot work from multiple processes because the `SyftObjectRegistry` is only in-memory cache, and couldn't load objects in a different process.
 - MongoDocumentStore is not serializable.
 - The ActionStore permissions are a bit unclear, and they need more further testing.
 - Race conditions not fixed yet.
 - The SQLite clients need to close the file on Windows(SQLiteActionStore, SQLiteStorePartition)

## Affected Dependencies
 - Added depends for testing: `pytest_mock_resources`, `pytest-lazy-fixture`, `python_on_whales`

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Comment by rasswanth-s** [comment]
Wohoo!! Top quality Testing

File: packages/syft/src/syft/core/node/new/action_store.py [hunk 1]
```diff
@@ -83,6 +83,15 @@ class ActionStore:
 
 @serializable(recursive_serde=True)
 class KeyValueActionStore(ActionStore):
+    """Generic Key-Value Action store.
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/node/new/action_store.py:86]: 🙌 Documentation

**Comment by madhavajay** [comment]
@bcebere I couldn't push to this PR again for some reason, I fixed some issues in this commit here: https://github.com/OpenMined/PySyft/pull/7443/commits/3566eb808bde824b1949a65c8c478df653d62af5

**Comment by bcebere** [comment]
> @bcebere I couldn't push to this PR again for some reason, I fixed some issues in this commit here: [3566eb8](https://github.com/OpenMined/PySyft/commit/3566eb808bde824b1949a65c8c478df653d62af5)

OK, sure. But I didn't touch those files in this PR.

File: packages/syft/setup.cfg [hunk 2]
```diff
@@ -92,7 +93,11 @@ test_plugins =
     pytest-asyncio
     pytest-randomly
     pytest-sugar
+    pytest_mock_resources
+    python_on_whales
```
Comment (madhavajay) [inline packages/syft/setup.cfg:97]: this looks really handy

File: packages/syft/src/syft/core/node/new/action_store.py (hunk 1 above)
Comment (madhavajay) [reply packages/syft/src/syft/core/node/new/action_store.py:86]: 💯

File: packages/syft/src/syft/core/node/new/document_store.py [hunk 3]
```diff
@@ -294,9 +312,14 @@ def __init__(
...
         self.init_store()
 
-    def init_store(self) -> None:
-        self.unique_cks = self.settings.unique_keys.all
-        self.searchable_cks = self.settings.searchable_keys.all
+    def init_store(self) -> Result[Ok, Err]:
+        try:
+            self.unique_cks = self.settings.unique_keys.all
+            self.searchable_cks = self.settings.searchable_keys.all
+        except BaseException as e:
```
Comment (madhavajay) [inline packages/syft/src/syft/core/node/new/document_store.py:319]: I think in a seperate task we can write some code which can automatically run a function, capture any exceptions and propagate them back as an Err (with some information about where it was caught) or Ok if everything is fine so we can get some clarity on where exceptions occur.

File: packages/syft/src/syft/core/node/new/sqlite_document_store.py [hunk 4]
```diff
@@ -68,9 +82,15 @@ def _connect(self) -> None:
...
         self.file_path = self.store_config.client_config.file_path
-        self._db[thread_ident()] = sqlite3.connect(self.file_path)
+        self._db[thread_ident()] = sqlite3.connect(
+            self.file_path,
+            timeout=self.store_config.client_config.timeout,
+            check_same_thread=self.store_config.client_config.check_same_thread,
+        )
+
+        # Set journal mode to WAL.
+        self._db[thread_ident()].execute("pragma journal_mode=wal")
```
Comment (madhavajay) [inline packages/syft/src/syft/core/node/new/sqlite_document_store.py:93]: nice, WAL should be helpful

File: packages/syft/src/syft/core/node/new/syft_object.py [hunk 5]
```diff
@@ -144,6 +145,11 @@ def make_id(cls, values: Dict[str, Any]) -> Dict[str, Any]:
     __attr_repr_cols__: List[str] = []  # show these in html repr collections
 
     def to_mongo(self) -> Dict[str, Any]:
+        warnings.warn(
```
Comment (madhavajay) [inline packages/syft/src/syft/core/node/new/syft_object.py:148]: awesome

File: packages/syft/tests/syft/locks_test.py [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+
+
+@pytest.mark.parametrize(
+    "config",
+    [
+        pytest.lazy_fixture("locks_file_config"),
+        pytest.lazy_fixture("locks_redis_config"),
+    ],
+)
+@pytest.mark.skipif(sys.platform != "linux", reason="Testing Mongo only on Linux")
```
Comment (madhavajay) [inline packages/syft/tests/syft/locks_test.py:221]: Any reason we only test this on linux?

**Review by madhavajay — APPROVED** [review]
Definitely PR of the year so far! 👏

**Pull request merged**